lib/tic_tac_toe_game.rb

```diff
@@ -75,46 +73,50 @@ def advance_to_next_player
     end
   end
 
-  def reset_current_player_pointer
-    @current_player_pointer = 0
-  end
-
-  def advance_current_player_pointer
-    @current_player_pointer += 1 
-  end
-
-  def current_player
-    players[current_player_pointer]
-  end
-
   def handle_game_over
     display_game_over_message
     ask_user_to_play_again
   end
 
   def display_game_over_message
     if rules.game_won?(board)
-      user_interface.render_message_game_won_by(current_player)
+      render(user_view.game_won_by(current_player))
     elsif rules.game_tied?(board)
-      user_interface.render_message_game_tied
+      render(user_view.game_tied)
     end
   end
 
   def ask_user_to_play_again
-    user_interface.render_message_play_again?
+    render(user_view.request_user_play_again)
     player_selection = gets.chomp
     if player_selection == "y" || player_selection == "Y"
       reset_game
     else
-      user_interface.render_good_bye
+      render(user_view.good_bye)
     end
   end
 
   def reset_game
     board.clear_board
     reset_current_player_pointer
-    user_interface.render_board_with_numbers
+    render(user_view.board_with_numbers)
     play_game
   end
 
+  def render(message)
+    user_output_stream.render(message)
+  end
+
+  def current_player
+    players[current_player_pointer]
+  end
+  
+  def reset_current_player_pointer
+    @current_player_pointer = 0
+  end
+
+  def advance_current_player_pointer
+    @current_player_pointer += 1 
+  end
```

Comment:
Could you find a way to rearrange the `players` array without having to maintain this state? Maybe you could ensure that the current player is always `players.first` and the rest are in order.

Reply:
Done through pairing.